cilium/cmd: improve 'bpf metrics list' JSON output

jibi opened the pull request:
This commit changes the JSON output for the 'bpf metrics list' command from the current format:

```js
{
  "reason:0 dir:1": [
    "count:63752 bytes:20913476"
  ],
  "reason:0 dir:2": [
    "count:29697 bytes:2489416"
  ],
  "reason:132 dir:2": [
    "count:27 bytes:1930"
  ]
}
```

which is hard to parse even with tools such as jq to a more structured format:

```js
[
  {
    "reason": 0,
    "description": "Success",
    "values": {
      "egress": {
        "packets": 29921,
        "bytes": 2508163
      },
      "ingress": {
        "packets": 64220,
        "bytes": 21064892
      }
    }
  },
  {
    "reason": 132,
    "description": "Invalid source ip",
    "values": {
      "egress": {
        "packets": 27,
        "bytes": 1930
      }
    }
  }
]
```

jibi committed:
mockmaps: add support for metricsmap

Signed-off-by: Gilberto Bertin <gilberto@isovalent.com>

kkourt commented on cilium/cmd/bpf_metrics_list.go:
```diff
@@ -37,31 +38,85 @@ const (
+func listMetrics(m metricsmap.MetricsMap) {
+	bpfMetricsList := make(map[string]string)
+	callback := func(key bpf.MapKey, value bpf.MapValue) {
+		bpfMetricsList[key.String()] = value.String()
```
This is a minor thing, but AFAICT here we transform the numbers into strings only to transform them back into numbers for the JSON case. Would it make sense to have a different callback function for the JSON case so that we don't we do the double transformation? I think the resulting code might end up being simpler.

jibi replied:
Here we are still dealing with entries which have keys like `reason:0 dir:1` and values like `count:63752 bytes:20913476` (see [here](https://github.com/cilium/cilium/blob/master/pkg/maps/metricsmap/metricsmap.go#L104-L122) for the definition of `func (vs Values) String() string` and `func (k Key) String() string`) which need to first get split into the individual values so I don't think there's much we can optimize yet

kkourt replied:
Thanks for the explanation!

It wasn't so much about improving performance but rather to potentially simplify things. Seems this is not as easy as I originally thought.

kkourt reviewed (commented):
Overall, LGTM.

One thing I wonder is whether we are breaking any existing user scripts with this change.

rolinh reviewed (commented):
This is undoubtedly an improvement! The parsing is a little too fragile for my taste but there's unit test coverage and it seems to do the job.

However, the output differs from the suggested format in #13492. Would you mind explaining the rationale for choosing a map with the reason number as string for keys rather than a list of objects which include the reason as in #13492?

christarazi commented on cilium/cmd/bpf_metrics_list.go:
```diff
@@ -129,6 +165,44 @@ func listMetrics(bpfMetricsList map[string][]string) {
+func extractRow(key, value string) (*metricsRow, error) {
+	reasonCodeStr, directionCodeStr, ok := extractTwoValues(key)
+	if !ok {
+		return nil, fmt.Errorf("Cannot extract reason and traffic direction from map's key \"%s\"", key)
```
Nit: error strings should start lower-cased

Source: https://github.com/golang/go/wiki/CodeReviewComments#error-strings

christarazi commented on cilium/cmd/bpf_metrics_list.go:
```diff
@@ -129,6 +165,44 @@ func listMetrics(bpfMetricsList map[string][]string) {
+		return nil, fmt.Errorf("Cannot parse bytes counter: %s", err)
+	}
+
+	return &metricsRow{reasonCode, reasonDesc, direction, packets, bytes}, nil
```
Style nit, feel free to ignore. I just reordered the statements that don't return an error together so that the pattern of the code is consistent with functions that return error, to functions that don't return error.

```suggestion
	directionCode, err := strconv.Atoi(directionCodeStr)
	if err != nil {
		return nil, fmt.Errorf("Cannot parse direction: %s", err)
	}

	packetsStr, bytesStr, ok := extractTwoValues(value)
	if !ok {
		return nil, fmt.Errorf("Cannot extract packets and bytes counters from map's value \"%s\"", value)
	}

	packets, err := strconv.Atoi(packetsStr)
	if err != nil {
		return nil, fmt.Errorf("Cannot parse packets counter: %s", err)
	}

	bytes, err := strconv.Atoi(bytesStr)
	if err != nil {
		return nil, fmt.Errorf("Cannot parse bytes counter: %s", err)
	}
	
	reasonDesc := monitorAPI.DropReason(uint8(reasonCode))
	direction := metricsmap.MetricDirection(uint8(directionCode))

	return &metricsRow{reasonCode, reasonDesc, direction, packets, bytes}, nil
```

christarazi commented on cilium/cmd/bpf_metrics_list_test.go:
```diff
@@ -0,0 +1,101 @@
+	. "gopkg.in/check.v1"
+)
+
+type BPFmetricsmapuite struct{}
```
```suggestion
type BPFMetricsMapSuite struct{}
```

christarazi commented on pkg/maps/metricsmap/metricsmap.go:
```diff
@@ -93,6 +93,19 @@ type Value struct {
 // Values is a slice of Values
 type Values []Value
 
+// A "Record" designates a map entry (key + value).
```
```suggestion
// Record designates a map entry (key + value).
```

christarazi reviewed (changes requested):
Output looks good to me, I like the list. A few comments below.

christarazi commented on cilium/cmd/bpf_metrics_list_test.go:
```diff
@@ -0,0 +1,110 @@
+
+var _ = Suite(&BPFMetricsmapuite{})
+
+func (s *BPFMetricsmapuite) TestDumpMetrics(c *C) {
```
This is still left over: https://github.com/cilium/cilium/pull/13731#discussion_r521580547

jibi replied:
:man_facepalming: :man_facepalming: 
time for larger terminal fonts I guess (sorry for the noise, just updated the PR)

jibi committed:
cilium/cmd: improve 'bpf metrics list' JSON output

This commit changes the JSON output for the 'bpf metrics list'
command from the current format:

```js
{
  "reason:0 dir:1": [
    "count:63752 bytes:20913476"
  ],
  "reason:0 dir:2": [
    "count:29697 bytes:2489416"
  ],
  "reason:132 dir:2": [
    "count:27 bytes:1930"
  ]
}
```

which is hard to parse even with tools such as jq to a more structured
format:

```js
[
  {
    "reason": 0,
    "description": "Success",
    "values": {
      "egress": {
        "packets": 29921,
        "bytes": 2508163
      },
      "ingress": {
        "packets": 64220,
        "bytes": 21064892
      }
    }
  },
  {
    "reason": 132,
    "description": "Invalid source ip",
    "values": {
      "egress": {
        "packets": 27,
        "bytes": 1930
      }
    }
  }
]
```

Fixes: #13492

Signed-off-by: Gilberto Bertin <gilberto@isovalent.com>

qmonnet reviewed (approved):
Thanks a lot!

rolinh reviewed (approved):
:rocket:
